File: android/app/src/main/java/com/linphonetest/LinphoneModule.kt
package com.linphonetest
import android.net.Uri
import android.os.Build
import android.telecom.CallAttributes
import android.util.JsonWriter
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.core.telecom.CallAttributesCompat
import androidx.core.telecom.CallsManager
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.bridge.WritableMap
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.supervisorScope

import org.linphone.core.*

class LinphoneModule(reactContext: ReactApplicationContext): ReactContextBaseJavaModule(reactContext) {
    override fun getName() = "LinphoneModule"

    private val TAG = "LinphoneModule"
    private var core: Core = Factory.instance().createCore(null, null, reactApplicationContext)

    private val coreListener: CoreListener = object : CoreListenerStub( ) {
        override fun onCallStateChanged(
            core: Core,
            call: Call,
            state: Call.State?,
            message: String
        ) {
            Log.d(TAG, "onCallStateChanged: $state $message")

            val params = Arguments.createMap().apply {
                putString("state", state.toString())
                putString("message", message)
            }
            sendEvent(reactContext, "callstate", params)

            when (state) {
                Call.State.IncomingReceived -> {

                }

                Call.State.Idle -> TODO()
                Call.State.PushIncomingReceived -> TODO()
                Call.State.OutgoingInit -> TODO()
                Call.State.OutgoingProgress -> TODO()
                Call.State.OutgoingRinging -> TODO()
                Call.State.OutgoingEarlyMedia -> TODO()
                Call.State.Connected -> TODO()
                Call.State.StreamsRunning -> TODO()
                Call.State.Pausing -> TODO()
                Call.State.Paused -> TODO()
                Call.State.Resuming -> TODO()
                Call.State.Referred -> TODO()
                Call.State.Error -> TODO()
                Call.State.End -> TODO()
                Call.State.PausedByRemote -> TODO()
                Call.State.UpdatedByRemote -> TODO()
                Call.State.IncomingEarlyMedia -> TODO()
                Call.State.Updating -> TODO()
                Call.State.Released -> TODO()
                Call.State.EarlyUpdatedByRemote -> TODO()
                Call.State.EarlyUpdating -> TODO()
                null -> TODO()
            }
        }
    }

    /**
     * React Native Event Emitter
     * */

    private var listenerCount = 0;

    private fun sendEvent(reactContext: ReactApplicationContext, eventName: String, params: WritableMap) {
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit(eventName, params)
    }

    @ReactMethod fun addListener(_eventName: String) {
        if( listenerCount == 0) core.addListener(coreListener)
        listenerCount += 1;
    }

    @ReactMethod fun removeListener() {
        listenerCount -= 1;
        if( listenerCount == 0) core.removeListener(coreListener)
    }

    @ReactMethod fun removeListeners() {
        listenerCount = 0;
        core.removeListener(coreListener)
    }

    /**
     * End of React native Event Emitter
     * */

    @ReactMethod fun register(username: String, password: String, domain: String, transport: String, promise: Promise) {
        val authInfo = Factory.instance().createAuthInfo(username, null, password, null, null, null);
        val accountParams = core.createAccountParams();

        val identity = Factory.instance().createAddress("sip:${username}@${domain}")
        accountParams.identityAddress = identity;

        val address = Factory.instance().createAddress("sip:${domain}")

        address?.transport = TransportType.valueOf(transport)
        accountParams.serverAddress = address
        accountParams.isRegisterEnabled = true

        val account = core.createAccount(accountParams)
        core.addAuthInfo(authInfo)
        core.addAccount(account)
        core.defaultAccount = account

        val listener = { _: Account, state: RegistrationState, message: String ->
            if (state !== RegistrationState.Ok && state !== RegistrationState.Progress) {
                promise.reject(state.toString(), message)
            } else if (state == RegistrationState.Ok) {
                promise.resolve("Registration successful")
            }
        }

        account.addListener(listener)
        core.start()
    }

    @ReactMethod fun unregister(promise: Promise) {
        val account = core.defaultAccount
        account ?: return

        val params = account.params
        val clonedParams = params.clone()

        clonedParams.isRegisterEnabled = false
        account.params = clonedParams

        account.addListener{ _, state, message ->
            if (state !== RegistrationState.Cleared && state !== RegistrationState.Progress) {
                promise.reject(state.toString(), message)
            } else if (state == RegistrationState.Cleared) {
                promise.resolve("Unregister successful")
            }
        }
    }

    @ReactMethod fun delete(promise: Promise) {
        val account = core.defaultAccount
        account ?: return

        core.removeAccount(account)

        promise.resolve("Delete successful")
    }

    @ReactMethod fun accept() {
        core.currentCall?.accept()
    }

    @ReactMethod fun terminate() {
        core.currentCall?.terminate()
    }

    @ReactMethod fun decline() {
        core.currentCall?.decline(Reason.Declined)
    }

    @ReactMethod fun call(address: String, promise: Promise) {
        val callParams = core.createCallParams(null)
        callParams ?: return promise.reject("call-creation", "Call params creation failed")

        callParams.mediaEncryption = MediaEncryption.SRTP

        val remoteAddress = Factory.instance().createAddress(address)
        remoteAddress ?: return promise.reject("call-creation", "Address creation failed")

        val call = core.inviteAddressWithParams(remoteAddress, callParams)

        if (call == null) {
            promise.reject("call-creation", "Call invite failed")
        } else {
            promise.resolve("Call successful")
        }
    }

    /**
     * Resolve um objeto com os dispositivos de áudio disponíveis e o id do dispositivo selecionado atualmente
     * */
    @ReactMethod fun getAudioDevices(promise: Promise){
        val values = Arguments.createArray()

        for (device in core.audioDevices) {
            val mappedDevice = Arguments.createMap()
            mappedDevice.putString("name", device.deviceName)
            mappedDevice.putString("driverName", device.driverName)
            mappedDevice.putString("id", device.id)
            mappedDevice.putString("type", device.type.toString())
            mappedDevice.putString("capabilities", device.capabilities.toString())
            values.pushMap(mappedDevice)
        }

        val returnedMap = Arguments.createMap()
        returnedMap.putArray("devices", values)
        returnedMap.putString("current", core.currentCall?.outputAudioDevice?.id)

        promise.resolve(returnedMap)
    }

    @ReactMethod fun setAudioDevice(id: String, promise: Promise) {
        if (core.currentCall == null) {
            promise.reject("no-call", "No current call")
            return
        }

        val newDevice = core.audioDevices.find{ it.id == id }
        if (newDevice == null) {
            promise.reject("no-device", "No device with $id found")
            return
        }
        if (newDevice.id == core.currentCall?.outputAudioDevice?.id) {
            return
        }

        core.currentCall!!.outputAudioDevice = core.audioDevices.find { it.id == id }
        promise.resolve("Set audio device successfully")
    }
}

File: ios/LinphoneModule.swift
//
//  LinphoneModule.swift
//  linphoneTest
//
//  Created by Mateus Mello on 02/02/24.
//

import Foundation

import linphonesw
import os

@available(iOS 14.0, *)
@objc(LinphoneModule)
class LinphoneModule: RCTEventEmitter {
  
  var core: Core!
  var account: Account?
  var coreDelegate: CoreDelegate!
  
  private var listenersCount: Int = 0
  let logger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "main")
  
  override init() {
    LoggingService.Instance.logLevel = LogLevel.Debug
    
    try? core = Factory.Instance.createCore(configPath: "", factoryConfigPath: "", systemContext: nil)
    try? core.start()
    
    super.init()
  }
  
  override class func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  /**
   * React native Event Emitter
   */
  override func supportedEvents() -> [String]! {
    return ["callstate"]
  }
  
  override func startObserving() {
    coreDelegate = CoreDelegateStub(
      
      onCallStateChanged: { (core: Core, call: Call, state: Call.State, message: String) in
        self.sendEvent(withName: "callstate", body: [
          "message": message,
          "state": String(describing: state)
        ])
      }
    )
    
    core.addDelegate(delegate: coreDelegate)
  }
  
  override func stopObserving() {
    core.removeDelegate(delegate: coreDelegate)
  }
  /**
   * End of React Native Event Emitter
   */
  
  @objc
  public func register(_ _username: String, password _password: String,domain _domain: String,transport _transport: String, resolver _resolver: @escaping RCTPromiseResolveBlock, rejecter _rejecter: @escaping RCTPromiseRejectBlock) {
    
    let registrationDelegate = AccountDelegateStub(onRegistrationStateChanged: {
      (account: Account, state: RegistrationState, message: String) in
      if (state != .Ok && state != .Progress) {
        _rejecter(String(describing: state), message, NSError(domain: "register", code: state.rawValue))
      } else if (state == .Ok) {
        _resolver("Registration successful")
      }
    })
    
    do {
      let transport = switch _transport {
        case "Tls": TransportType.Tls
        case "Tcp": TransportType.Tcp
        default: TransportType.Udp
      }
      
      let authInfo = try Factory.Instance.createAuthInfo(username: _username, userid: nil, passwd: _password, ha1: "", realm: "", domain: _domain)
      let accountParams = try core.createAccountParams()
      let identity = try Factory.Instance.createAddress(addr: "sip:\(_username)@\(_domain)")
      let address = try Factory.Instance.createAddress(addr: "sip:\(_domain)")
      
      try accountParams.setIdentityaddress(newValue: identity)
      try address.setTransport(newValue: transport)
      try accountParams.setServeraddress(newValue: address)
      
      accountParams.registerEnabled = true
      account = try core.createAccount(params: accountParams)
      core.addAuthInfo(info: authInfo)
      try core.addAccount(account: account!)
      core.defaultAccount = account
            
      account?.addDelegate(delegate: registrationDelegate)
    } catch {
      logger.error("registration error \(error.localizedDescription)")
      account?.removeDelegate(delegate: registrationDelegate)
      _rejecter("register", error.localizedDescription, error)
    }
  }
  
  @objc public func unregister(_ resolver: @escaping RCTPromiseResolveBlock, reject rejecter: @escaping RCTPromiseRejectBlock) {
    if let account = core.defaultAccount {
      let clonedParams = account.params?.clone()
      clonedParams?.registerEnabled = false
      account.params = clonedParams
      
      let registrationDelegate = AccountDelegateStub(onRegistrationStateChanged: {
        (account: Account, state: RegistrationState, message: String) in
        if (state == .Ok) {
            resolver("Unregistration successful")
        } else {
          rejecter(String(describing: state), message, NSError(domain: "unregister", code: state.rawValue))
        }
      })
      
      account.addDelegate(delegate: registrationDelegate)
    }
  }
  
  @objc public func deleteAccount() {
    if let account = core.defaultAccount {
      core.removeAccount(account: account)
      core.clearAccounts()
      core.clearAllAuthInfo()
    }
  }
  
  @objc public func accept() {
      try? core.currentCall?.accept()
  }
  
  @objc public func terminate() {
    try? core.currentCall?.terminate()
  }
  
  @objc public func decline() {
    try? core.currentCall?.decline(reason: Reason.Declined)
  }
  
  @objc(call: resolve: reject:)
  public func call(_ address: String, resolve resolver: RCTPromiseResolveBlock, reject rejecter: RCTPromiseRejectBlock) {
    logger.info("Calling \(address)")
    
    do {
      let remoteAddress = try Factory.Instance.createAddress(addr: address)
      let callParams = try core.createCallParams(call: nil)
      
      callParams.mediaEncryption = MediaEncryption.None
      let _ = core.inviteAddressWithParams(addr: remoteAddress, params: callParams)
      
      resolver(true)
    } catch {
      logger.error("Error Calling: \(error.localizedDescription)")
      rejecter("call-creation", error.localizedDescription, error)
    }
  }
  
  @objc public func getAudioDevices(_ resolver: RCTPromiseResolveBlock, reject rejecter: RCTPromiseRejectBlock) {
    var values: [[String:String]] = []
    
    for device in core.audioDevices {
      
      let capabilities = {
        switch (device.capabilities) {
        case .CapabilityPlay:
          return "CapabilityPlay"
        case .CapabilityRecord:
          return "CapabilityRecord"
        default:
          return "CapabilityAll"
        }
      }()
      
      let mappedDevice: [String:String] = [
        "name": device.deviceName,
        "driverName": device.driverName,
        "id": device.id,
        "type": String(describing: device.type),
        "capabilities": capabilities
      ]
      
      values.append(mappedDevice)
    }
    
    let returned: [String: Any] = [
      "devices": values,
      "current": core.currentCall?.outputAudioDevice?.id ?? ""
    ]
    resolver(returned)
  }
  
  @objc public func setAudioDevice(_ id: String, resolve resolver: RCTPromiseResolveBlock, reject rejecter: RCTPromiseRejectBlock) {
    if (core.currentCall == nil) {
      rejecter("no-call", "no current call", NSError(domain: "call", code: 1))
    }
    
    if let newDevice = core.audioDevices.first(where: { (i) in i.id == id }) {
      if (newDevice.id == core.currentCall?.outputAudioDevice?.id) {
        return
      }
      
      core.currentCall?.outputAudioDevice = newDevice
    } else {
      rejecter("no-device", "No device with \(id) found", NSError(domain: "no-device", code: 0))
    }
  }
}
